presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -584,6 +585,25 @@ public PlanWithProperties visitExplainAnalyze(ExplainAnalyzeNode node, Preferred
             return rebaseAndDeriveProperties(node, child);
         }
 
+        @Override
+        public PlanWithProperties visitAnalyzeFinish(AnalyzeFinishNode node, PreferredProperties context)
+        {
+            PlanWithProperties child = planChild(node, PreferredProperties.any());
+
+            // if the child is already a gathering exchange, don't add another
+            if ((child.getNode() instanceof ExchangeNode) && ((ExchangeNode) child.getNode()).getType().equals(GATHER)) {
+                return rebaseAndDeriveProperties(node, child);
```

Comment:
Analyze Finish is supposed to run on coordinator, here we don't know if this GATHER will be run on the coordinator.

Also, this means that final aggregation will be run on the coordinator. We save some network bandwidth for analyze results, but we end up running final aggregation on the coordinator. Am i correct? Should we worry about overwhelming the coordinator?

What if we remove this `if` altogether?

Reply:
There is double gather. You can check the planner test. The final aggregations runs on worker, and then `TableFInishOperator` runs on coordinator. The `coordinator only` property is set in `PropertyDerivations`.